### PR TITLE
WIP: implement union.

### PR DESCRIPTION
**Pull request by csukuangfj**
TODOs

- [ ] support arc_map
- [ ] compute `row_splits` of the output fsa in the lambda

### CODE REVIEW

**Comment by csukuangfj** [comment]
Since k2 supports visualizing only a single FSA,  one possible use of `union` is  convert a FsaVec to a single FSA and then
visualize the converted FSA.

File: k2/csrc/fsa_algo.h [hunk 1]
```diff
@@ -239,6 +239,19 @@ FsaVec LinearFsas(Ragged<int32_t> &symbols);
 Ragged<int32_t> ShortestPath(FsaVec &fsas,
                              const Array1<int32_t> &entering_arcs);
 
+/* Compute the union of all fsas in a FsaVec.
+
```
Comment (danpovey) [inline k2/csrc/fsa_algo.h:243]: OK... this is not the most general interface that we might need, but could be OK for now.
We might eventually need to do a union (FsaVec, FsaVec -> FsaVec), or (FsaVec, Fsa -> FsaVec);
or in fact (FsaVec, [Ragged<Int> that shows how the Fsas in the FsaVec are to be grouped]) -> FsaVec.

But we can add other version later or extend the interface.

File: k2/csrc/fsa_algo.cu [hunk 2]
```diff
@@ -540,4 +540,93 @@ void AddEpsilonSelfLoops(FsaOrVec &src, FsaOrVec *dest,
...
+  K2_LOG(INFO) << out_arcs;
+  K2_LOG(INFO) << out_row_ids;
+
+  // TODO(fangjun): figure out out_row_splits in the lambda
+  //
+  // RaggedShape shape = RaggedShape2(&out_row_splits, &out_row_ids,
+  // num_out_arcs);
+  RaggedShape shape = RaggedShape2(nullptr, &out_row_ids, num_out_arcs);
+  Fsa ans = Ragged<Arc>(shape, out_arcs);
+  return ans;
```
Comment (danpovey) [inline k2/csrc/fsa_algo.cu:629]: If you are not going to support the arc-map for this PR, this should probably be documented and you should assert that it's nullptr.

Comment (csukuangfj) [reply k2/csrc/fsa_algo.cu:629]: I am going to support arc-map in this PR. Not finished yet.

File: k2/csrc/fsa_algo_test.cu [hunk 3]
```diff
@@ -376,4 +374,26 @@ TEST(FsaAlgo, ShortestPath) {
...
+  )";
+
+  for (auto &context : {GetCpuContext(), GetCudaContext()}) {
+    Fsa fsa1 = FsaFromString(s1);
+    Fsa fsa2 = FsaFromString(s2);
+    Fsa *fsa_array[] = {&fsa1, &fsa2};
+    FsaVec fsa_vec = CreateFsaVec(2, &fsa_array[0]);
+    fsa_vec = fsa_vec.To(context);
+    Fsa fsa = Union(fsa_vec);
+    K2_LOG(INFO) << FsaToString(fsa);
```
Comment (qindazhu) [inline k2/csrc/fsa_algo_test.cu:395]: Again, I **strongly** suggest we add random tests (especially empty case) for any algorithms, it really will take our too much time to dubug in the final python training scripts if there's any issue. You can use RandomFsa/RandomFsaVec to create Fsas.

File: k2/csrc/fsa_algo.cu [hunk 4]
```diff
@@ -540,4 +540,93 @@ void AddEpsilonSelfLoops(FsaOrVec &src, FsaOrVec *dest,
...
+  const int32_t *fsas_row_ids2_data = fsas.RowIds(2).Data();
+  const Arc *arcs_data = fsas.values.Data();
+
+  int32_t num_fsas = fsas.Dim0();
+  int32_t num_states = fsas.TotSize(1);
+  int32_t num_arcs = fsas.TotSize(2);
+
+  // A new start state and a new final state are added (+2).
+  // The final state of each fsa is removed (-num_fsas)
+  int32_t num_out_states = num_states + 2 - num_fsas;
```
Comment (qindazhu) [inline k2/csrc/fsa_algo.cu:560]: what happened here if one of input Fsa is empty?

File: k2/csrc/fsa_algo.cu [hunk 5]
```diff
@@ -540,4 +540,93 @@ void AddEpsilonSelfLoops(FsaOrVec &src, FsaOrVec *dest,
...
+  int32_t out_final_state = num_out_states - 1;
+
+  // For every fsa, a new arc is added from the new start state
+  // to its original start state (+num_fsas)
+  int32_t num_out_arcs = num_arcs + num_fsas;
```
Comment (qindazhu) [inline k2/csrc/fsa_algo.cu:565]: and here, if one of the input fsa is empty. That's why I strongly ask you add random tests and I think I have asked this too many times..

**Comment by danpovey** [comment]
Mm, search for lambda_set_fsa_nonempty in fsa_algo.cu for an example of
code that deals with the empty FSA issue.

**Comment by danpovey** [comment]
For dealing with empty FSAs, we might be able to solve it in a general way,
like:

  /*
     Remove empty FSAs from `src`, setting `dest` to an FsaVec containing
FSAs in the
     same order but with empty FSAs removed.
        ...
  */
  RemoveEmptyFsas(FsaVec &src, Renumbering *renumbering, FsaVec *dest);

That would be sufficient in this case.

**Comment by qindazhu** [comment]
> RemoveEmptyFsas(FsaVec &src, Renumbering *renumbering, FsaVec *dest);

I like this approach to deal with empty case in a general way, but I just doubt it's not efficient in some cases, e.g the original algorithm is relatively simple or even has only one kernel in it, call `RemoveEmptyFsas` will add 2 kernel calls before it.

For simple algorithms, I would like to call `RemoveEmptyFsas` before we feed those fsas into a chain of operations, e,g,
```
RemoveEmptyFsas();
union()
rmepsilon()
determinize()
...
```

**Comment by danpovey** [comment]
Mm.  Most algorithms are slow enough that we wouldn't notice the
difference.  I'm concerned about adding extra assumptions that are not
quickly
checkable.  Unless we make the properties an extra argument to these
function calls.  Since we are currently computing the properties whenever
we construct an Fsa, this might be feasible.

In the majority of cases, there won't be empty FSAs; so we could have a
renumbering inside each algorithm that only happens if
there were empty FSAs, and that wouldn't really impact the speed.  So
regardless, I don't see a good reason to add "nonempty" as a
requirement for algorithms (but we can consider making the input properties
an extra arg to functions).

**Comment by qindazhu** [comment]
> I don't see a good reason to add "nonempty" as a
requirement for algorithms (but we can consider making the input properties
an extra arg to functions).

Agree

File: k2/csrc/fsa_algo.cu (hunk 4 above)
Comment (csukuangfj) [reply k2/csrc/fsa_algo.cu:560]: It will abort the program.

It indeed checks for the empty case
https://github.com/k2-fsa/k2/blob/c1e16fc0ea65b5f7be322291b6852d465b44637b/k2/csrc/fsa_algo.cu#L581


The empty case is not supported as the precondition is documented clearly
in the comment  of this function
https://github.com/k2-fsa/k2/blob/c1e16fc0ea65b5f7be322291b6852d465b44637b/k2/csrc/fsa_algo.h#L244-L245

File: k2/csrc/fsa_algo.cu [hunk 6]
```diff
@@ -540,4 +540,93 @@ void AddEpsilonSelfLoops(FsaOrVec &src, FsaOrVec *dest,
...
+
+    if (arc.dest_state == this_fsa_final_state_idx1)
+      arc.dest_state = out_final_state;
+    else
+      arc.dest_state = arc.dest_state - arc.src_state + out_state_idx0;
+
+    arc.src_state = out_state_idx0;
+    out_arcs_data[out_arc_idx01] = arc;
+
+    if (fsas_state_idx1 == 0 && fsas_arc_idx2 == 0) {
```
Comment (qindazhu) [inline k2/csrc/fsa_algo.cu:612]: it's start state of the origianl Fsa, right?  we can just use `fsas_state_idx0x == fsa_state_idx01`

File: k2/csrc/fsa_algo.cu [hunk 7]
```diff
@@ -540,4 +540,93 @@ void AddEpsilonSelfLoops(FsaOrVec &src, FsaOrVec *dest,
...
+      // add a new arc from the new start state to this state
```
Comment (qindazhu) [inline k2/csrc/fsa_algo.cu:613]: and may be it's better to change `this_state` to something like `start state of this fsa`? If I understand correctly

File: k2/csrc/fsa_algo.cu (hunk 4 above)
Comment (qindazhu) [reply k2/csrc/fsa_algo.cu:560]: Oh, seems I miss your documentation here, sorry. But as @danpovey said below, I now also think we should not non=empty requirements on such basic algorithms, we should handle such cases in it.

**Comment by danpovey** [comment]
I'm OK with the current implementation if it checks for nonempty and also
documents it.
We can fix this later, possibly along with making properties an arg to
functions.
Let's just get this merged quickly so Yiming (@freewym) can make progress.

**Comment by qindazhu** [comment]
OK

**Comment by csukuangfj** [comment]
Will support arc map later today 



Sent from myMail for iOS


Sunday, 15 November 2020, 12:19 +0800 from notifications@github.com  <notifications@github.com>:

**Comment by qindazhu** [comment]
> Will support arc map later today 

As @freewym does not need arc_map stuff for now, I think you can add arc_map support (with row_splits) in another PR.

**Comment by danpovey** [comment]
OK I'll merge this now.

**Comment by csukuangfj** [comment]
Not wrong but it's the 0th arc of the start state.



Sent from myMail for iOS


Sunday, 15 November 2020, 12:11 +0800 from notifications@github.com  <notifications@github.com>:

File: k2/csrc/fsa_algo_test.cu (hunk 3 above)
Comment (csukuangfj) [reply k2/csrc/fsa_algo_test.cu:395]: Will add it.
